### ISSUE TITLE
False Negative: Dead stores are missed once they appear inside more complex control flow and resource-management structures.

### ISSUE THREAD

**Issue by Carlson-JLQ** [pmd/pmd#6540]
# False Negative: Dead stores are missed once they appear inside more complex control flow and resource-management structures.

Version
7.22.0

## Checker
- Checker id: `category/java/bestpractices.xml/UnusedAssignment`
- Checker description: This rule detects assignments to variables where the assigned value is never read before being overwritten or going out of scope.

## Description of the false negative
`diff/output` marks 3 false negatives for `category/java/bestpractices.xml/UnusedAssignment`. Across these samples, the benchmark is still exercising the same underlying bug pattern, but the implementation appears to lose track of it once the source deviates from one canonical shape. The misses read more like matching gaps than deliberate exclusions.

### `PosCase2.java`
Source: `output/pmd/UnusedAssignment/src/main/java/scensct/core/pos/PosCase2.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: A local variable updated via postfix increment where the old value is never read should be flagged as unused assignment.

```java
// A local variable updated via postfix increment where the old value is never read should be flagged as unused assignment.
package scensct.core.pos;

public class PosCase2 {
    public static void main(String[] args) {
        int x = 5;
        int unused = x++; // Postfix increment result (old value 5) never read.
        System.out.println(x); // Reading new value, not the old one.
    }
}
```

### `PosCase2.java`
Source: `output/sonarqube/DeadStoreCheck/src/main/java/scensct/core/pos/PosCase2.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: A method with a block body declares a local variable with an explicit initializer, and the variable is never read or used later before it goes out of scope or is reassigned, without try-finally or resources.

```java
// A method with a block body declares a local variable with an explicit initializer, and the variable is never read or used later before it goes out of scope or is reassigned, without try-finally or resources.
package scensct.core.pos;

public class PosCase2 {
    public void deadInitializer() {
        int y = 10; // Explicit initializer, never read later // [REPORTED LINE]
        System.out.println("Done");
    }
}
```

### `PosCase6.java`
Source: `output/sonarqube/DeadStoreCheck/src/main/java/scensct/core/pos/PosCase6.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: A method with a block body declares a local variable with an explicit initializer, and the variable is never read or used later before it goes out of scope or is reassigned, with try-finally but no local variable used in finally.

```java
// A method with a block body declares a local variable with an explicit initializer, and the variable is never read or used later before it goes out of scope or is reassigned, with try-finally but no local variable used in finally.
package scensct.core.pos;

public class PosCase6 {
    public void deadInitializerWithTryFinally() {
        int y = 10; // Explicit initializer, never read later // [REPORTED LINE]
        try {
            System.out.println("In try");
        } finally {
            System.out.println("Finally");
        }
    }
}
```

## Cause analysis
The rule is documented as follows: This rule detects assignments to variables where the assigned value is never read before being overwritten or going out of scope.

 The missed cases suggest that the implementation is narrower than the rule description and too dependent on one preferred AST shape.

Once the same behavior is expressed through a helper, an overload, a modifier such as `abstract` or `native`, a different comment position, or a slightly rearranged control-flow structure, the report disappears even though the benchmark still treats the case as in-scope.

For `category/java/bestpractices.xml/UnusedAssignment`, the practical improvement would be to generalize the match so it follows the underlying behavior rather than one exact spelling of it.

## References
None known.

**Comment by oowekyala** [comment]
Unused variables are not reported by default because of the overlap with UnusedLocalVariable.
